File: lua/skill/NfComboSkill.lua
require 'global'

local NfComboSkill = {
    skill,
    attacker,
    attackers
}

function NfComboSkill.onBegin(self)
    local skill = self.skill

    self.attackers = skill:GetComboCharList()    

    skill:HideScene()    

    NfComboSkill.PlayCombAnimAndEffect(self)

    skill:StartCameraAnim()

    startCoroutine(function( )
        NfComboSkill.Attack(self)
    end)
end

function NfComboSkill.PlayCombAnimAndEffect(self)
    local skill = self.skill
    local attackers = self.attackers

    for i=0, attackers.Count-1 do
        local c = attackers[i]
        c.position = Vector3.zero
        c.rotation = c.SrcRotation
        c:PlayAnim("heji")
    end
    skill:PlayEffect(self.attacker, skill.FireEffect, 3)
end

function NfComboSkill.OnStopCameraAnim(self)
    self.skill:ShowScene()

    NfComboSkill.MoveSrc(self)
end

function NfComboSkill.Attack( self )
    local skill = self.skill
    local attacker = self.attacker

    local time = skill.TotalTime
    if time < 0.01 then
        time = attacker:GetAnimLength("heji")
    end

    local dam = skill:GetComboDamage(self.attackers)

    local bulletModel = skill.BulletModel
    local hitTime = skill.HitTime

    skill:AddEvent(hitTime, function()
        local endTime = time - hitTime
        local targets = skill:FindTargets(true)

        for i=0, targets.Count-1 do
            local c = targets[i]
            local beginPos = attacker.position + Vector3(0,0.6,0)
            local endPos = Vector3(c.position.x, 0.6, c.position.z)
            skill:PlayChainEff(bulletModel, endTime, beginPos, endPos)
            skill:CalcDamage(attacker, c, dam, 0)
        end

        skill:End(endTime)
    end)
end

function NfComboSkill.onEnd(self)
    NfComboSkill.MoveSrc(self)

    self.skill:DestroyAllEff()
end

function NfComboSkill.MoveSrc(self)
    local attackers = self.attackers
    for i=0, attackers.Count-1 do
        local c = attackers[i]
        c.position = c.SrcPos
        c:Idle()
    end
end

return NfComboSkill

File: Skills/NfComboSkill.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using LitJson;

public class NfComboSkill : NfSkill
{
    protected List<Character> attackers;

    protected override void OnSkillBegin()
    {
        attackers = GetComboCharList();

        HideScene();

        PlayCombAnimAndEffect();
        StartCameraAnim();

        StartCoroutine(ComboAtk());
    }

    protected override void OnSkillEnd()
    {
        for (int i = 0; i < attackers.Count; ++i)
        {
            Character c = attackers[i];
            c.position = c.SrcPos;
            c.Idle();
        }

        DestroyAll();
        DestroyAllEff();
    }

    protected override void StopCameraAnim()
    {
        base.StopCameraAnim();

        for (int i = 0; i < attackers.Count; ++i)
        {
            Character c = attackers[i];
            c.position = c.SrcPos;
            c.Idle();
        }

        ShowScene();
    }

    protected virtual IEnumerator ComboAtk()
    {
        float time = TotalTime;
        if (time < 0.01f)
        {
            time = animation["heji"].length;
        }

        int dam = GetComboDamage(attackers);

        string bulletModel = BulletModel;
        float hitTime = HitTime;

        List<Character> targets = FindTargets(true);

        yield return new WaitForSeconds(hitTime);

        for (int i = 0; i < targets.Count; ++i)
        {
            Character c = targets[i];

            Vector3 beginPos = attacker.position + new Vector3(0f, 0.6f, 0f);
            Vector3 endPos = new Vector3(c.position.x, 0.6f, c.position.z);
            PlayChainEff(bulletModel, time - hitTime, beginPos, endPos);
            CalcDamage(attacker, c, dam, 0f);
        }

        yield return new WaitForSeconds(time - hitTime);

        End();
    }

    protected virtual void PlayCombAnimAndEffect()
    {
        for (int i = 0; i < attackers.Count; ++i)
        {
            Character c = attackers[i];
            //if (attacker.camp == CampType.Friend)
            //    c.position = Fight.Inst.GetSlotPos(attacker.camp, 11);
            //else
            c.position = Vector3.zero;
            c.rotation = c.SrcRotation;
            c.PlayAnim("heji");
        }
        PlayEffect(attacker.gameObject, FireEffect, 3f);
    }
}
